### PR TITLE
[Core] add an option to compile Ray with jemalloc

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

add bazel rule to build jemalloc. we can then use `bazel build //:gcs_server --define jemalloc=true` to build with jemalloc.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by mwtian — APPROVED** [review]
Do we plan to build wheels with jemalloc?

File: bazel/BUILD.jemalloc [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  export CXX=$$(absolutize $(CC))
+  [ "$$(uname)" = "Linux" ] && export AR=$$(absolutize $(AR))
+  export NM=$$(absolutize $(NM))
+  export OBJCOPY=$$(absolutize $(OBJCOPY))
+  export CFLAGS=$(CC_FLAGS)
+  export CXXFLAGS=$(CC_FLAGS)
+  # export LTOFLAGS="-flto=thin" # todo: on next clang toolchain upgrade, check if it's fixed and we can re-enable thinlto on mac
+  export LTOFLAGS=""
+  export EXTRA_CFLAGS="$${LTOFLAGS}"
+  export EXTRA_CXXFLAGS="-libstdc++ $${EXTRA_CFLAGS}"
```
Comment (mwtian) [inline bazel/BUILD.jemalloc:36]: Why is this needed? It seems `-libstdc++` will not be useful with clang.

**Comment by rkooo567** [comment]
Is it still WIP?

File: src/ray/util/jemalloc_test.cc [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    size_t epoch = 1;
+    size_t previous_allocated = 0;
+    for (auto i = 0ul; i < 10; ++i) {
+        void* ptr = 0;
+        ASSERT_EQ(0, posix_memalign(&ptr, 16, 100000));
+        size_t sz, allocated, active, metadata, resident, mapped;
+        sz = sizeof(size_t);
+        mallctl("thread.tcache.flush", NULL, NULL, NULL, 0);
+        mallctl("epoch", NULL, NULL, &epoch, sizeof(epoch));
+        if (mallctl("stats.allocated", &allocated, &sz, NULL, 0) == 0
```
Comment (clarng) [inline src/ray/util/jemalloc_test.cc:28]: why 'if' and not 'assert'?

File: bazel/ray_deps_setup.bzl [hunk 3]
```diff
@@ -319,3 +319,11 @@ def ray_deps_setup():
         # When you first run this tool, it'll recommend a sha256 hash to put here with a message like: "DEBUG: Rule 'hedron_compile_commands' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = ..."
         sha256 = "7fbbbc05c112c44e9b406612e6a7a7f4789a6918d7aacefef4c35c105286930c",
     )
+
+    auto_http_archive(
```
Comment (rkooo567) [inline bazel/ray_deps_setup.bzl:323]: Don't we need a sanity PR to build a wheel with this and run? (or is it supposed to be a follow up?)
